### DISCUSSION TITLE
how to get specific user for login

### DISCUSSION THREAD

**Discussion by hanbu97** [maxcountryman/axum-login#29] [Q&A]
https://github.com/maxcountryman/axum-login/blob/0b87c32c344ade8eff9c9926545c00c3f84551e4/examples/sqlite/src/main.rs#L55

Here, can only search user=1  from the database. If I want to login some user else, what should I do.

Can you add a more general example? Thank you very much.

**Comment by maxcountryman** [comment]
This will largely depend on your application.

For instance, one common way you might acquire a potential user is through an HTML form, where someone attempting to log in provides a username and password.

Your application would then do whatever validation of the user your use case deems sufficient (e.g. this could be validating the username and password in some way, perhaps via your database). Having done so, you would then retrieve an instance of your user type (the same you implemented `AuthUser` for) with the specific details of this user.

That said, this is only one way applications might choose to operate and by design, `axum-login` doesn't make any assumptions about how you'll authenticate or store the user which is attempting to log in.

**Reply by Weijun-H** [reply]
Maybe we could add a straightforward example of how to get a specific user.

**Reply by maxcountryman** [reply]
That's possible. If you want to try, one way to do this would be to setup a SQLite db with seeded users and then write a demo application that uses it.

**Comment by hanbu97** [comment]
Thank you. Understood.

Another thing is, if we use something like bcrypt
``` rust
fn get_password_hash(&self) -> String {
    let hashed_pwd = bcrypt::hash(&self.password, DEFAULT_COST).unwrap();
    hashed_pwd
}
```
The thing is the hash is changing every time, but can be verified by bcrypt::verify(input_passwd, stored_passwd). 

How to deal with it.

**Comment by maxcountryman** [comment]
I would strongly advise storing the hashed password and not the password. If you do, then this method just returns that stored hash and should not repeat the hashing process.

What I've done using similar hashing algos (Argon2 for instance) is then check the provided password verifies with the stored hash. So for your bcrypt example, that crate should provide some method for checking a given password against a hash.
